integrations/access/common/access_monitoring_rules.go

```diff
@@ -0,0 +1,182 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package common
+
+import (
+	"context"
+	"maps"
+	"sync"
+
+	"github.com/gravitational/trace"
+
+	accessmonitoringrulesv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessmonitoringrules/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/integrations/access/common/teleport"
+	"github.com/gravitational/teleport/integrations/lib/logger"
+)
+
+const (
+	// defaultAccessMonitoringRulePageSize is the default number of rules to retrieve per request
+	defaultAccessMonitoringRulePageSize = 10
+)
+
+type AccessMonitoringRuleHandler struct {
+	accessMonitoringRules AmrMap
+
+	apiClient  teleport.Client
+	pluginType string
+
+	ruleAppliesCallback        func(amr *accessmonitoringrulesv1.AccessMonitoringRule) bool
+	fetchRecipientCallback     func(ctx context.Context, recipient string) (*Recipient, error)
+	matchAccessRequestCallback func(expr string, req types.AccessRequest) (bool, error)
+}
+
+// AmrMap is a concurrent map for access monitoring rules.
+type AmrMap struct {
+	sync.RWMutex
+	// rules are the access monitoring rules being stored.
+	rules map[string]*accessmonitoringrulesv1.AccessMonitoringRule
+}
+
+type AccessMonitoringRuleHandlerConfig struct {
+	Client     teleport.Client
+	PluginType string
+
+	RuleAppliesCallback        func(amr *accessmonitoringrulesv1.AccessMonitoringRule) bool
```

Comment:
IS there a use case for allowing custom callbacks? If the "rule applies" logic is always checking the plugin anme, can we remove this field and use the plugin type instead?